crates/chia-sdk-driver/src/primitives/datalayer/datastore.rs

```diff
@@ -0,0 +1,2278 @@
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct DataStore<M = DataStoreMetadata> {
+    /// The coin that holds this ``DataStore``.
+    pub coin: Coin,
+    /// The lineage proof for the singletonlayer.
+    pub proof: Proof,
+    /// The info associated with the ``DataStore``, including the metadata.
+    pub info: DataStoreInfo<M>,
+}
+
+impl<M> DataStore<M>
+where
+    M: ToClvm<Allocator> + FromClvm<Allocator>,
+{
+    pub fn new(coin: Coin, proof: Proof, info: DataStoreInfo<M>) -> Self {
+        DataStore { coin, proof, info }
+    }
+
+    /// Creates a coin spend for this ``DataStore``.
+    pub fn spend(self, ctx: &mut SpendContext, inner_spend: Spend) -> Result<CoinSpend, DriverError>
+    where
+        M: ToClvm<Allocator> + FromClvm<Allocator> + Clone,
+    {
+        let (puzzle_ptr, solution_ptr) = if self.info.delegated_puzzles.is_empty() {
+            let layers = self
+                .info
+                .clone()
+                .into_layers_without_delegation_layer(inner_spend.puzzle);
+
+            let solution_ptr = layers.construct_solution(
+                ctx,
+                SingletonSolution {
+                    lineage_proof: self.proof,
+                    amount: self.coin.amount,
+                    inner_solution: NftStateLayerSolution {
+                        inner_solution: inner_spend.solution,
+                    },
+                },
+            )?;
+
+            (layers.construct_puzzle(ctx)?, solution_ptr)
+        } else {
+            let layers = self.info.clone().into_layers_with_delegation_layer(ctx)?;
+            let puzzle_ptr = layers.construct_puzzle(ctx)?;
+
+            let delegated_puzzle_hash = tree_hash(&ctx.allocator, inner_spend.puzzle);
+
+            let tree = get_merkle_tree(ctx, self.info.delegated_puzzles)?;
+
+            let inner_solution = DelegationLayerSolution {
+                // if running owner puzzle, the line below will return 'None', thus ensuring correct puzzle behavior
+                merkle_proof: tree.get_proof(delegated_puzzle_hash.into()),
+                puzzle_reveal: inner_spend.puzzle,
+                puzzle_solution: inner_spend.solution,
+            };
+
+            let solution_ptr = layers.construct_solution(
+                ctx,
+                SingletonSolution {
+                    lineage_proof: self.proof,
+                    amount: self.coin.amount,
+                    inner_solution: NftStateLayerSolution { inner_solution },
+                },
+            )?;
+            (puzzle_ptr, solution_ptr)
+        };
+
+        let puzzle = ctx.serialize(&puzzle_ptr)?;
+        let solution = ctx.serialize(&solution_ptr)?;
+
+        Ok(CoinSpend::new(self.coin, puzzle, solution))
+    }
+
+    /// Returns the lineage proof that would be used by the child.
+    pub fn child_lineage_proof(&self, ctx: &mut SpendContext) -> Result<LineageProof, DriverError>
+    where
+        M: ToTreeHash,
```

Comment:
We shouldn't require `ToTreeHash` if we're using the `ctx` anyways. I'm thinking we should just get rid of all of the `ToTreeHash` bounds for metadata and use `ctx` to calculate the hash with `tree_hash`. Since metadata should generally be `NodePtr`, `Program`, or a metadata type that can be converted to a `NodePtr` anyways.